Archive/Imo/Imo1998Q2.lean

```diff
@@ -0,0 +1,265 @@
+
+/-- The second judge from an ordered pair of judges. -/
+abbrev JudgePair.judge₂ : JudgePair J → J :=
+  Prod.snd
+#align imo1998_q2.judge_pair.judge₂ Imo1998Q2.JudgePair.judge₂
+
+/-- The proposition that the judges in an ordered pair are distinct. -/
+abbrev JudgePair.Distinct (p : JudgePair J) :=
+  p.judge₁ ≠ p.judge₂
+#align imo1998_q2.judge_pair.distinct Imo1998Q2.JudgePair.Distinct
+
+/-- The proposition that the judges in an ordered pair agree about a contestant's rating. -/
+abbrev JudgePair.Agree (p : JudgePair J) (c : C) :=
+  r c p.judge₁ ↔ r c p.judge₂
+#align imo1998_q2.judge_pair.agree Imo1998Q2.JudgePair.Agree
+
+/-- The contestant from the triple consisting of a contestant and an ordered pair of judges. -/
+abbrev AgreedTriple.contestant : AgreedTriple C J → C :=
+  Prod.fst
+#align imo1998_q2.agreed_triple.contestant Imo1998Q2.AgreedTriple.contestant
+
+/-- The ordered pair of judges from the triple consisting of a contestant and an ordered pair of
+judges. -/
+abbrev AgreedTriple.judgePair : AgreedTriple C J → JudgePair J :=
+  Prod.snd
+#align imo1998_q2.agreed_triple.judge_pair Imo1998Q2.AgreedTriple.judgePair
+
+@[simp]
+theorem JudgePair.agree_iff_same_rating (p : JudgePair J) (c : C) :
+    p.Agree r c ↔ (r c p.judge₁ ↔ r c p.judge₂) :=
+  Iff.rfl
+#align imo1998_q2.judge_pair.agree_iff_same_rating Imo1998Q2.JudgePair.agree_iff_same_rating
+
+/-- The set of contestants on which two judges agree. -/
+def agreedContestants [Fintype C] (p : JudgePair J) : Finset C :=
+  Finset.univ.filter fun c => p.Agree r c
+#align imo1998_q2.agreed_contestants Imo1998Q2.agreedContestants
+section
+
+variable [Fintype J] [Fintype C]
+
+/-- All incidences of agreement. -/
+def A : Finset (AgreedTriple C J) :=
+  Finset.univ.filter @fun (a : AgreedTriple C J) =>
+    (a.judgePair.Agree r a.contestant ∧ a.judgePair.Distinct)
+#align imo1998_q2.A Imo1998Q2.A
+
+theorem A_maps_to_offDiag_judgePair (a : AgreedTriple C J) :
+    a ∈ A r → a.judgePair ∈ Finset.offDiag (@Finset.univ J _) := by simp [A, Finset.mem_offDiag]
+#align imo1998_q2.A_maps_to_off_diag_judge_pair Imo1998Q2.A_maps_to_offDiag_judgePair
+
+theorem A_fibre_over_contestant (c : C) :
+    (Finset.univ.filter fun p : JudgePair J => p.Agree r c ∧ p.Distinct) =
+      ((A r).filter fun a : AgreedTriple C J => a.contestant = c).image Prod.snd := by
+  ext p
+  simp only [A, Finset.mem_univ, Finset.mem_filter, Finset.mem_image, true_and_iff, exists_prop]
+  constructor
+  · rintro ⟨h₁, h₂⟩; refine' ⟨(c, p), _⟩; tauto
+  · intro h; aesop
+#align imo1998_q2.A_fibre_over_contestant Imo1998Q2.A_fibre_over_contestant
+
+theorem A_fibre_over_contestant_card (c : C) :
+    (Finset.univ.filter fun p : JudgePair J => p.Agree r c ∧ p.Distinct).card =
+      ((A r).filter fun a : AgreedTriple C J => a.contestant = c).card := by
+  rw [A_fibre_over_contestant r]
+  apply Finset.card_image_of_injOn
+  -- porting note: used to be `tidy`
+  unfold Set.InjOn; intros; ext; all_goals aesop
```

Comment:
Can you just add `extCore` as an extra tactic to `aesop`, like in the category theory library?

Reply:
That sounds like a good idea!

So I tried adding
```
-- We turn on `ext` inside `aesop`.
attribute [aesop unsafe] Std.Tactic.Ext.extCore
```
but even doing this in right in this file, didn't have the desired effect that I can leave out `ext` in the proof.

What would be the right file to add `extCore` to aesop? In the file where `extCore` is defined doesn't seem to be a good place, because I can't seem to import `Aesop` from there.

Reply:
You need to do it like in `CategoryTheory.Basic`. i.e. call `extCore'` with the correct signature, otherwise it treats it as an apply lemma, I think, rather than a tactic.

Reply:
I know it's a bit annoying setting up automation, but in the long run it's a good idea. :-)

Reply:
In particular, it might make sense to move `extCore'` to somewhere that can be imported by both this file and CategoryTheory.Basic.

Possibly even as high at `Ext` in Std4 itself!

Reply:
ok, I opened two draft pull requests:
https://github.com/leanprover/std4/pull/161
and 
https://github.com/leanprover-community/mathlib4/pull/5329

Applying both it seems that mathlib4 is able to build. However, I'm still not quite sure how to make it work in the Imo1998Q2 file.

Reply:
I think it would be best to move this not into std but rather somewhere where you can apply the aesop attribute in the same file.

Reply:
What file would be a good place then?

Reply:
This is not fixed in the sense that it still doesn't work. However, there is a porting note explaining what to do, so from my point of view this can be done in a separate pull request. (As soon as I know where to put it and get it working...)

Reply:
Yeah, let's merge this (but see my other suggestion) and clean up when the issue is resolved.

bors d+